conduit/pipeline.go

```diff
@@ -0,0 +1,299 @@
+	Exporter   nameConfigPair   `yaml:"Exporter"`
+}
+
+// Valid validates pipeline config
+func (cfg *PipelineConfig) Valid() error {
+	if cfg.ConduitConfig == nil {
+		return fmt.Errorf("PipelineConfig.Valid(): conduit configuration was nil")
+	}
+
+	if _, err := log.ParseLevel(cfg.pipelineLogLevel); err != nil {
+		return fmt.Errorf("PipelineConfig.Valid(): pipeline log level (%s) was invalid: %w", cfg.pipelineLogLevel, err)
+	}
+
+	if len(cfg.Importer.Config) == 0 {
+		return fmt.Errorf("PipelineConfig.Valid(): importer configuration was empty")
+	}
+
+	if len(cfg.Exporter.Config) == 0 {
+		return fmt.Errorf("PipelineConfig.Valid(): exporter configuration was empty")
+	}
+
+	return nil
+}
+
+// MakePipelineConfig creates a pipeline configuration
+func MakePipelineConfig(logger *log.Logger, cfg *Config) (*PipelineConfig, error) {
+	if cfg == nil {
+		return nil, fmt.Errorf("MakePipelineConfig(): empty conduit config")
+	}
+
+	// double check that it is valid
+	if err := cfg.Valid(); err != nil {
+		return nil, fmt.Errorf("MakePipelineConfig(): %w", err)
+	}
+
+	pCfg := PipelineConfig{pipelineLogLevel: logger.Level.String(), ConduitConfig: cfg}
+
+	// Search for pipeline configuration in data directory
+	autoloadParamConfigPath := filepath.Join(cfg.ConduitDataDir, autoLoadParameterConfigName)
+
+	_, err := os.Stat(autoloadParamConfigPath)
+	paramConfigFound := err == nil
+
+	if !paramConfigFound {
+		return nil, fmt.Errorf("MakePipelineConfig(): could not find %s in data directory (%s)", autoLoadParameterConfigName, cfg.ConduitDataDir)
+	}
+
+	logger.Infof("Auto-loading Conduit Configuration: %s", autoloadParamConfigPath)
+
+	file, err := os.Open(autoloadParamConfigPath)
+
+	err = viper.ReadConfig(file)
+	if err != nil {
+		return nil, fmt.Errorf("MakePipelineConfig(): reading config error: %w", err)
+	}
+
+	err = viper.Unmarshal(&pCfg)
+	if err != nil {
+		return nil, fmt.Errorf("MakePipelineConfig(): config file (%s) was mal-formed yaml: %w", autoloadParamConfigPath, err)
+	}
+
+	if err := pCfg.Valid(); err != nil {
+		return nil, fmt.Errorf("MakePipelineConfig(): config file (%s) had mal-formed schema: %w", autoloadParamConfigPath, err)
+	}
+
+	pCfg.PipelineLogLevel, err = log.ParseLevel(pCfg.pipelineLogLevel)
+	if err != nil {
+		// Belt and suspenders.  Valid() should have caught this
+		return nil, fmt.Errorf("MakePipelineConfig(): config file (%s) had mal-formed log level: %w", autoloadParamConfigPath, err)
+	}
+
+	return &pCfg, nil
+
+}
+
+// Pipeline is a struct that orchestrates the entire
+// sequence of events, taking in importers, processors and
+// exporters and generating the result
+type Pipeline interface {
+	Start() error
+	Stop() error
+}
+
+type pipelineImpl struct {
+	ctx    context.Context
+	cfg    *PipelineConfig
+	logger *log.Logger
+
+	initProvider *data.InitProvider
+
+	importer   *importers.Importer
+	processors []*processors.Processor
+	exporter   *exporters.Exporter
+}
+
+func (p *pipelineImpl) Start() error {
```

Comment:
consider renaming to `initialize`, since I'm sure there is going to be a lot more to do for `Start` in the next iteration.